### PR TITLE
ADR 0000: Architectural Decision Records

### LINKED ISSUE #2924: Start using ADRs to document architectural decisions
Instead of having internal design documents, we should start using [Architectural Decision Records](https://adr.github.io/) to document architectural decisions in this repository (under `docs/adr`). This is similar to our previous RFC process, but has existing tooling available.

### PR DESCRIPTION
**Pull request by kostko**
Fixes #2924 

Introduce architectural decision records (ADRs) for keeping track of
architecture decisions in a transparent way.

### CODE REVIEW

File: docs/adr/template.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+- {date}: {changelog}
+
+## Status
+
+> A decision may be "proposed" if it hasn't been agreed upon yet, or "accepted"
+> once it is agreed upon. If a later ADR changes or reverses a decision, it may
+> be marked as "deprecated" or "superseded" with a reference to its replacement.
+
+{Deprecated|Proposed|Accepted|Rejected}
```
Comment (ptrus) [inline docs/adr/template.md:13]: above comment also mentions `"superseded"`

File: docs/adr/template.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+> state, and why there is a problem. It should be as succinct as possible and
+> introduce the high level idea behind the solution.
+
+## Decision
+
+> This section explains all of the details of the proposed solution, including
+> implementation details. It should also describe affects / corollary items that
+> may need to be changed as a part of this. If the proposed change will be
+> large, please also indicate a way to do the change to maximize ease of review
+> (e.g. the optimal split of things to do between separate PR's).
```
Comment (tjanez) [inline docs/adr/template.md:27]:
```suggestion
> (e.g. the optimal split of things to do between separate pull requests (PRs)).
```

File: docs/adr/template.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+### Positive
+
+### Negative
+
+### Neutral
+
+## References
+
+> Are there any relevant PR comments, issues that led up to this, or articles
+> referenced for why we made the given design choice? If so link them here!
```
Comment (tjanez) [inline docs/adr/template.md:43]:
```suggestion
> referenced for why we made the given design choice? If so, link them here!
```

File: docs/adr/index.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+* `<adr-number>` is the assigned zero-padded ADR number.
+* `<short-title>` is the ADR's title in [Kebab case].
+
+The content of each ADR should follow [the template]. In short, an ADR should
+provide:
+
+* a changelog of all modifications so far,
+* context on the relevant goals and the current state,
+* proposed changes to achieve the goals,
+* summary of pros and cons and
```
Comment (tjanez) [inline docs/adr/index.md:31]: Perhaps also explicitly state that an ADR should also include the decision:

```suggestion
* proposed changes to achieve the goals,
* decision was made,
* summary of pros and cons and
```

File: docs/adr/index.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  committers. After a sufficient amount of discussion, acceptance or rejection
+  decision will be taken in accoordance with the [governance process] and the
+  pull request will be merged, introducing a new ADR.
+
+After the ADR is merged an implementation may be undertaken by following the
+[contribution process].
+
+[your branch]: ../../CONTRIBUTING.md#contributing-code
+[changelog fragment]: ../../.changelog/README.md
+[governance process]: ../../GOVERNANCE.md
```
Comment (tjanez) [inline docs/adr/index.md:66]: The Governance process doc doesn't exist yet. Perhaps omit a link until we have one? Or link to an issue to add it in the mean time?

Comment (kostko) [reply docs/adr/index.md]: Removed the link.

File: docs/adr/index.md (hunk 4 above)
Comment (kostko) [reply docs/adr/index.md]: Good idea, done.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -39,24 +39,20 @@ most effective at receiving input and making progress.
...
-repository. Such issues will typically receive `p:3` priority in their initial
-triage.
+repository.
 
 If the feature is **more complicated**, involves protocol changes, or has
-potential safety or performance implications, then first check with us whether
-it is something we are willing to consider and is in-scope for our project.
+potential safety or performance implications, then consider [proposing an ADR]
+and submit it as a pull request ot this repository. This will allow a structured
+review and commenting of the proposed changes. You should aim to reach consensus
```
Comment (peterjgilbert) [inline CONTRIBUTING.md:47]: Consensus among who? Any committers or contributors involved in the discussion?

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -157,24 +154,25 @@ https://github.com/oasisprotocol/oasis-core/blob/master/README.md).
...
 
 Documentation is always welcome! Documentation comes in several forms:
 
 * Code-level documentation, following language specifications.
-* Mechanism documentation, which live in `docs/` describes commonly used
-  systems, like adding a new protocol definition, or benchmarking the system.
-* Protocol documentation, living in the [RFCs repository](
-  https://github.com/oasislabs/rfcs), describes protocols and architectural
-  design.
+* Developer and system documentation, which lives in `docs/` describes commonly
```
Comment (peterjgilbert) [inline CONTRIBUTING.md:174]:
```suggestion
* Developer and system documentation, which lives in `docs/`, describes commonly
```

File: docs/adr/index.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  pull request will be merged, introducing a new ADR.
+
+After the ADR is merged an implementation may be undertaken by following the
+[contribution process].
+
+[your branch]: ../../CONTRIBUTING.md#contributing-code
+[changelog fragment]: ../../.changelog/README.md
+[contribution process]: ../../CONTRIBUTING.md
+
+## Current Records
```
Comment (peterjgilbert) [inline docs/adr/index.md:69]: Do we plan on updating this list every time we add an adr?

Comment (kostko) [reply docs/adr/index.md:69]: Yes, it's described in the process.

File: CONTRIBUTING.md (hunk 6 above)
Comment (kostko) [reply CONTRIBUTING.md:47]: This was there before, I guess consensus in general, the actual decision making will be made more clear in the governance document (which is not there yet). I'll do some rewording to make this more clear.

Comment (kostko) [reply CONTRIBUTING.md:47]: Replaced with:
> You should aim to get the ADR accepted and merged before starting on implementation.

**Pull request merged**